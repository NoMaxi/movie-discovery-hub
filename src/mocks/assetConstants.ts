// Mock version of `src/constants/assetConstants.ts` for tests (mock images imports)
export const DEFAULT_IMAGE_URL = "test-image-stub";
