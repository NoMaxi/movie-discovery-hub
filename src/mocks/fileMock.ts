// Stub for image file imports in Jest tests
export default "";
