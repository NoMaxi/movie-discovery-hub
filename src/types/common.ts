export type Genre = "All" | "Documentary" | "Comedy" | "Horror" | "Crime";

export interface Movie {
    id: string;
    imageUrl: string;
    title: string;
    releaseYear: number;
    genres: string[];
}
