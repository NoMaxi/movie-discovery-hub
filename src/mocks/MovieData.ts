import { InitialMovieInfo, MovieDetailsData } from "@/types/common";

export const mockMovieDetails: MovieDetailsData = {
    id: "1",
    title: "Pulp Fiction",
    releaseYear: 1994,
    genres: ["Comedy", "Crime"],
    imageUrl: "https://upload.wikimedia.org/wikipedia/en/3/3b/Pulp_Fiction_%281994%29_poster.jpg",
    rating: 8.9,
    duration: 154,
    description:
        "Jules Winnfield (Samuel L. Jackson) and Vincent Vega (John Travolta) are two hit men who are out to retrieve a suitcase stolen from their employer, mob boss Marsellus Wallace (Ving Rhames). Wallace has also asked Vincent to take his wife Mia (Uma Thurman) out a few days later when Wallace himself will be out of town. Butch Coolidge (Bruce Willis) is an aging boxer who is paid by Wallace to lose his fight. The lives of these seemingly unrelated people are woven together comprising of a series of funny, bizarre and uncalled-for incidents.â€”Soumitra",
};

export const alternativeMovieDetails: MovieDetailsData = {
    id: "2",
    imageUrl: "https://example.com/poster-alt.jpg",
    title: "Another Great Film",
    releaseYear: 2022,
    genres: ["Comedy", "Crime"],
    rating: 7.9,
    duration: 59.5,
    description: "The greatest movie in history of cinema.",
};

export const mockMovieInfo: InitialMovieInfo = {
    ...mockMovieDetails,
    releaseDate: "1994-10-14",
};
