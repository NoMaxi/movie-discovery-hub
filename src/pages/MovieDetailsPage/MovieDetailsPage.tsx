import { useLoaderData } from "react-router-dom";
import { Movie } from "@/types/common";
import { CloseDetailsButton } from "@/components/common/CloseDetailsButton/CloseDetailsButton";
import { MovieDetails } from "@/components/MovieDetails/MovieDetails";

export const MovieDetailsPage = () => {
    const { movieDetails } = useLoaderData<{ movieDetails: Movie }>();

    return (
        <div data-testid="movie-details-container" className="relative">
            <CloseDetailsButton />
            <MovieDetails details={movieDetails} />
        </div>
    );
};
