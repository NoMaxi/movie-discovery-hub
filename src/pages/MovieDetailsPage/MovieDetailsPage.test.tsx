import { render, screen } from "@testing-library/react";
import { useLoaderData } from "react-router-dom";
import "@testing-library/jest-dom";
import { Movie } from "@/types/common";
import { mockAPIMovie } from "@/mocks/movieData";
import { MovieDetailsPage } from "./MovieDetailsPage";

jest.mock("react-router-dom", () => ({
    useLoaderData: jest.fn(),
}));

jest.mock("@/components/MovieDetails/MovieDetails", () => ({
    MovieDetails: ({ details }: { details: Movie }) => (
        <div data-testid="movie-details" data-details={JSON.stringify(details)} />
    ),
}));

jest.mock("@/components/common/CloseDetailsButton/CloseDetailsButton", () => ({
    CloseDetailsButton: () => <div data-testid="close-details-button" />,
}));

describe("MovieDetailsPage", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        (useLoaderData as jest.Mock).mockReturnValue({ movieDetails: mockAPIMovie });
    });

    test("Should render component correctly", () => {
        const { asFragment } = render(<MovieDetailsPage />);
        const mainContainer = screen.getByTestId("movie-details-container");

        expect(mainContainer).toHaveClass("relative");
        expect(screen.getByTestId("close-details-button")).toBeInTheDocument();
        expect(screen.getByTestId("movie-details")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should pass correct movie details to MovieDetails component", () => {
        render(<MovieDetailsPage />);
        const movieDetailsComponent = screen.getByTestId("movie-details");
        const passedDetails = JSON.parse(movieDetailsComponent.getAttribute("data-details") || "{}");

        expect(passedDetails).toEqual(mockAPIMovie);
    });

    test("Should handle different movie data correctly", () => {
        const differentMovieDetails = {
            ...mockAPIMovie,
            id: 456,
            title: "Another Movie",
            overview: "Another overview",
            poster_path: "https://example.com/another-poster.jpg",
            release_date: "2024-02-02",
            vote_average: 7.2,
        };
        (useLoaderData as jest.Mock).mockReturnValue({ movieDetails: differentMovieDetails });

        render(<MovieDetailsPage />);
        const movieDetailsComponent = screen.getByTestId("movie-details");
        const passedDetails = JSON.parse(movieDetailsComponent.getAttribute("data-details") || "{}");

        expect(passedDetails).toEqual(differentMovieDetails);
    });
});
