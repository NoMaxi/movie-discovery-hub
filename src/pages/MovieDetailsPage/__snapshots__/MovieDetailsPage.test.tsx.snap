// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetailsPage Should render component correctly 1`] = `
<DocumentFragment>
  <div
    class="relative"
    data-testid="movie-details-container"
  >
    <div
      data-testid="close-details-button"
    />
    <div
      data-details="{"id":123,"title":"Test Movie","poster_path":"https://example.com/poster.jpg","release_date":"2022-05-15","genres":["Action","Drama"],"overview":"A test movie description","runtime":120,"vote_average":7.5}"
      data-testid="movie-details"
    />
  </div>
</DocumentFragment>
`;
